File: claude-code/mind/hooks/prompt_submit.py
#!/usr/bin/env python3
"""
UserPromptSubmit hook: update the Living Mind context before each
prompt.

Reads the user's prompt from stdin, calls the mind for proactive
context relevant to that prompt, and prints the result to stdout.
Claude Code injects stdout from UserPromptSubmit hooks into the
context immediately before the user turn.

This is much cheaper than SessionStart: a single mind_think call
with `reasoning_depth="fast"` (the default), so it adds maybe
200ms to the agent's pre-turn latency.
"""
import json
import os
import sys
import urllib.request
import urllib.error


def main() -> int:
    nexus_url = os.environ.get("NEXUS_URL", "http://localhost:7777").rstrip("/")
    secret = os.environ.get("NEXUS_SECRET", "")
    mind_id = os.environ.get("MIND_ID", "default")
    agent_id = os.environ.get("AGENT_ID", "claude-code")

    if not secret:
        return 0

    # Read the prompt from stdin (Claude Code passes it as JSON)
    try:
        stdin_data = sys.stdin.read()
        if stdin_data.strip():
            data = json.loads(stdin_data)
            user_prompt = data.get("prompt", "")[:500]  # cap to 500 chars
        else:
            return 0
    except (json.JSONDecodeError, KeyError):
        return 0

    if not user_prompt:
        return 0

    headers = {
        "Authorization": f"Bearer {secret}",
        "Content-Type": "application/json",
    }

    try:
        req = urllib.request.Request(
            f"{nexus_url}/v1/mind/think",
            headers=headers, method="POST",
            data=json.dumps({
                "mind_id": mind_id,
                "question": user_prompt,
                "context": {"agent_id": agent_id},
                "reasoning_depth": "fast",
            }).encode(),
        )
        with urllib.request.urlopen(req, timeout=8) as resp:
            body = json.loads(resp.read())
    except (urllib.error.URLError, json.JSONDecodeError, KeyError, TimeoutError):
        return 0

    sections: list = []

    # Proactive context the mind thinks is relevant
    items = body.get("proactive_context", []) or []
    if items:
        lines = ["## Living Mind: Relevant Context\n"]
        for item in items[:3]:
            lines.append(f"- [{item['type']}] {item['content']} (relevance {item['relevance']:.2f})")
        sections.append("\n".join(lines))

    # If the mind is asking back, surface the question
    cq = body.get("clarifying_question")
    if cq:
        sections.append(f"## Living Mind: Clarifying Question\n{cq}")

    if not sections:
        return 0

    print("\n\n".join(sections))
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception:
        sys.exit(0)

File: integrations/plugins/opencode-mind/hooks/prompt_submit.mjs
// prompt_submit.mjs — OpenCode hook that injects just-in-time
// context from the Living Mind before each user prompt.
import { setTimeout as sleep } from 'node:timers/promises';

const NEXUS_URL = process.env.NEXUS_URL || 'http://localhost:7777';
const NEXUS_SECRET = process.env.NEXUS_SECRET || '';
const MIND_ID = process.env.MIND_ID || 'default';
const AGENT_ID = process.env.AGENT_ID || 'opencode';

async function mindThink(question) {
  if (!NEXUS_SECRET) return null;
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), 5000);
    const r = await fetch(`${NEXUS_URL}/v1/mind/think`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${NEXUS_SECRET}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        mind_id: MIND_ID,
        question,
        context: { agent_id: AGENT_ID },
        reasoning_depth: 'fast',
      }),
      signal: ctrl.signal,
    });
    clearTimeout(t);
    if (!r.ok) return null;
    return await r.json();
  } catch (e) {
    return null;
  }
}

async function readStdin() {
  return new Promise((resolve) => {
    let data = '';
    process.stdin.on('data', (chunk) => (data += chunk));
    process.stdin.on('end', () => resolve(data));
    setTimeout(() => resolve(data), 200);
  });
}

async function main() {
  const stdin = await readStdin();
  let userPrompt = '';
  try {
    userPrompt = (JSON.parse(stdin).prompt || '').slice(0, 500);
  } catch (e) {
    // ignore — empty stdin
  }
  if (!userPrompt) return;

  const think = await mindThink(userPrompt);
  if (!think) return;

  const sections = [];

  const items = think.proactive_context || [];
  if (items.length > 0) {
    const lines = ['## Living Mind: Relevant Context\n'];
    for (const item of items.slice(0, 3)) {
      lines.push(`- [${item.type}] ${item.content} (relevance ${item.relevance.toFixed(2)})`);
    }
    sections.push(lines.join('\n'));
  }

  if (think.clarifying_question) {
    sections.push(`## Living Mind: Clarifying Question\n${think.clarifying_question}`);
  }

  if (sections.length > 0) {
    process.stdout.write(sections.join('\n\n') + '\n');
  }
}

main().catch(() => { /* never fail a prompt */ });
